store/orm/orm.go

```diff
@@ -369,12 +364,37 @@ func (orm *ORM) CreateServiceAgreement(sa *models.ServiceAgreement) error {
 	return orm.DB.Create(sa).Error
 }
 
-// UnscopedJobRunsWithStatus returns all JobRuns, including ones that have been
-// soft deleted, which have the passed statuses.
-func (orm *ORM) UnscopedJobRunsWithStatus(statuses ...models.RunStatus) ([]models.JobRun, error) {
-	runs := []models.JobRun{}
-	err := orm.Unscoped().preloadJobRuns().Where("status IN (?)", statuses).Find(&runs).Error
-	return runs, err
+// UnscopedJobRunsWithStatus passes all JobRuns to a callback, one by one,
+// including those that were soft deleted.
+func (orm *ORM) UnscopedJobRunsWithStatus(cb func(*models.JobRun) error, statuses ...models.RunStatus) error {
+	return batch(1000, func(offset, limit uint) (uint, error) {
+		var runIDs []string
+		err := orm.DB.Unscoped().
+			Table("job_runs").
+			Where("status IN (?)", statuses).
+			Order("status desc").
+			Limit(limit).
+			Offset(offset).
+			Pluck("ID", &runIDs).Error
+		if err != nil {
+			return 0, fmt.Errorf("error finding job ids %v", err)
+		}
+
+		for _, id := range runIDs {
+			var run models.JobRun
+			err := orm.Unscoped().preloadJobRuns().First(&run, "id = ?", id).Error
```

Comment:
I actually envisioned doing the `Pluck` outside the pagination batch, taking all IDs in memory and only paginating when retrieving the fully hydrated runs. It's a significant perf improvement at the expense of holding all IDs in memory, but I'm hoping that latter issue is an unrealistic limit (we can hold 100k IDs in memory??). This way would work regardless of the size.

```golang
		err := orm.DB.Unscoped().
			Table("job_runs").
			Where("status IN (?)", statuses).
			Order("status desc, id").
			Limit(limit).
			Offset(offset).
			Pluck("ID", &allIDs).Error

return batch(100, func(offset, limit uint) (uint, error) {
    err := orm.Unscoped().preloadJobRuns().Where(ids).Limit(limit).Offset(offset).Find(&runs).Error
....
cb(singleRun)
....
}
```

Just thought I would mention it.

Reply:
The challenge for that is that it's hard to predict what size of query will actually produce the problem, since it's highly dependent on the shape of the data. 1000 JobRuns with 1 task, might be fine, but the only 500 with 2 tasks, 10 with 5 tasks and lots of overrides and so on. I think it would be safest to leave it like this for now, which shouldn't bump into any unexpectedly large queries, look into this: #164807802 and then revisit.

Reply:
#164807802 = https://www.pivotaltracker.com/n/projects/2129823/stories/164807802